### DISCUSSION TITLE
AI support

### DISCUSSION THREAD

**Discussion by basil** [neanes/neanes#263] [Ideas]
The most obvious way to integrate AI with this project would be to train a model with the classical repertoire and use this to autocomplete formulæ based on the first few neumes during score entry. The autocompletion could take into account information like the current mode/scale, but this would not be useful for composition due to a lack of awareness of the underlying text.

A more challenging application would be to integrate AI for automated composition. If large language models can be used to generate plausible paragraphs of text, it seems plausible that the same technology could be applied to generate plausible compositions. Consider the case of a model trained with the classical repertoire attempting to generate a plausible composition for a Greek liturgical text. For the result to be plausible, the model would likely need to include information not only about the music but also about the accentuation patterns of the lyrics. Similarly, the input text may need to be tokenized from words into accented/unaccented syllables. Papa Ephraim's binary representation of 0 (unaccented) and 1 (accented) may not be sufficient, either—much of the musical rhetoric in a composition depends on the rhetoric of the text. For example, imperfect musical cadences often correlate with the end of a clause within a sentence, while perfect musical cadences often correlate with the end of a sentence. In other words, the model would likely need to account for both musical and textual semantics.

The most challenging application would be to generate plausible compositions in a novel target language, like Romanian or English. All of the above challenges apply as well as a new one: mapping the textual semantics of the original language to the textual semantics of the target language in the model. ChatGPT can do automated translation, so perhaps there is already a solution for this.

As an aside, AI support could be a possible way to monetize this software—retain the open-source version for basic usage and develop a proprietary SaaS-based AI extension for revenue generation. (I am not sure if the GPLv3 allows this, but the project could always be relicensed to a different open-source license if needed.) I am not sure how big the market for this is, but it is certainly something I would be willing to pay for.

**Comment by danielgarthur** [comment]
I think something like this would need to be streamlined. For example, you feed a program a PDF, it converts it to images and uses OCR to recognize the neumes and the text. Remove the "extra" melismatic vowels and detect half-words that are sometimes repeated in papadic and sticheraric hymns, then using something like [Morphologia Graeca ](https://latin-dict.github.io/dictionaries/morphology-grc.html) to determine which syllables are stressed.

You might check on Psaltologion, especially in the Greek section because I think something like this is already being undertaken (not the AI part, but the catalog of neumes and syllables), although I'm not sure whether it's publicly available.

For OCR, there is [Gamera](https://gamera.informatik.hsnr.de/addons/psaltiki/), but I've never used it. I think the author of Melodos made his own OCR algorithms.

**Comment by Mykbibby** [comment]
Let's keep AI out of this...

**Comment by danielgarthur** [comment]
I'm trying out Github Discussions. I'm going to move this Issue over to the Discussions sections and keep the Issues section full of concrete, actionable work, and the Discussions section more of a place for developing ideas that can eventually become actionable issues..

**Comment by basil** [comment]
> I'm going to […] keep the Issues section full of concrete, actionable work

Then why is https://github.com/danielgarthur/neanes/issues/81 still open?

**Comment by basil** [comment]
I suppose the first step in training a deep learning model is to define a tokenization scheme for Byzantine notation. [This page](https://miditok.readthedocs.io/en/latest/tokenizations.html) features a comparison of MIDI-based tokenization schemes. [This article](https://huggingface.co/blog/juancopi81/using-hugging-face-to-train-a-gpt-2-model-for-musi) demonstrates how to use the MMM tokenization scheme to convert MIDI files into a tokenized dataset and then train a GP2-model from the resulting symbolic representation.

Since Byzantine music is such a niche field, I suppose that all the Byzantine music ever composed in human history could fit into a few gigabytes of tokenized text. Considering that Microsoft’s Phi-2 [claims to have used](https://www.microsoft.com/en-us/research/blog/phi-2-the-surprising-power-of-small-language-models/) “14 days on 96 A100 GPUs” (which works out to around $35,000 [using current Lambda pricing](https://lambdalabs.com/service/gpu-cloud)) with 1.4T training tokens, it seems plausible that a Byzantine music model could be trained at a significant fraction of that cost and that one could use e.g. an Anastasimatarion to train a small-scale model on consumer workstation hardware for development purposes.

Altogether, I don't see any reason why a Byzantine music equivalent of [MuseNet](https://openai.com/research/musenet) or [Magenta's Music Transformer](https://magenta.tensorflow.org/music-transformer) couldn't be created using existing machine learning technology.
